### PR TITLE
Preliminary work to add Swap Panes functionality (GH Issues 1000, 4922)

### LINKED ISSUE #1000: Deliverable: Add support for panes
This is the megathread for tracking all the remaining work that needs to be done to get panes ready.

* [ ] Panes should be resizable with the keyboard.
* [ ] Panes should be resizable with the mouse. The user should be able to drag
  the separator for a pair of panes, and have the content between them resize as
  the separator moves.
* [ ] There's no keyboard shortcut for "ClosePane"
* [ ] The user should be able to configure what profile is used for splitting a
  pane. Currently, the default profile is used, but it's possible a user might
  want to create a new pane with the parent pane's profile.
* [ ] There should be some sort of UI to indicate that a particular pane is
  focused, more than just the blinking cursor. `tmux` accomplishes this by
  colorizing the separators adjacent to the active pane. Another idea is
  displaying a small outline around the focused pane (like when tabbing through
  controls on a webpage).
* [ ] The user should be able to navigate the focus of panes with the keyboard,
  instead of requiring the mouse.
* [ ] The user should be able to zoom a pane, to make the pane take the entire
  size of the terminal window temporarily.
* [ ] A pane doesn't necessarily need to host a terminal. It could potentially
  host another UIElement. One could imagine enabling a user to quickly open up a
  Browser pane to search for a particular string without needing to leave the
  terminal.
* [ ] Open panes with something other than the default profile
* [ ] If you click on the separator's between panes, then on panes will be focused
* [ ] Panes should have an optional motion effect on entrance
* [ ] Add a default keybinding for panes, to enable them by default

### PR DESCRIPTION
**Pull request by Rosefield**
## Summary of the Pull Request
Add functionality to swap a pane with an adjacent (Up/Down/Left/Right) neighbor.

 
## References
This work potentially touches on: #1000 #2398 and #4922

## PR Checklist
* [x] Closes a component of #1000 (partially, comment), #4922 (partially, `SwapPanes` function is added but not hooked up, no detach functionality)
* [x] CLA signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/Terminal) and sign the CLA
* [x] Tests added/passed
* [ ] Documentation updated. If checked, please file a pull request on [our docs repo](https://github.com/MicrosoftDocs/terminal) and link it here: #xxx
* [x] Schema updated.
* [ ] I've discussed this with core contributors already. If not checked, I'm ready to accept this work might be rejected in favor of a different grand plan. Issue number where discussion took place: #xxx

## Detailed Description of the Pull Request / Additional comments

Its been a while since I've written C++ code, and it is my first time working on a Windows application. I hope that I have not made too many mistakes.

Work currently done:
- Add boilerplate/infrastructure for argument parsing, hotkeys, event handling
- Adds the `MovePane` function that finds the focused pane, and then tries to find
  a pane that is visually adjacent to according to direction.
- First pass at the `SwapPanes` function that swaps the tree location of two panes
- First working version of helpers `_FindFocusAndNeighbor` and `_FindNeighborFromFocus`
  that search the tree for the currently focused pane, and then climbs back up the tree
  to try to find a sibling pane that is adjacent to it. 
- An `_IsAdjacent' function that tests whether two panes, given their relative offsets, are adjacent to each other according to the direction.

Next steps:
- Once working these functions (`_FindFocusAndNeighbor`, etc) could be utilized to also solve #2398 by updating the `NavigateFocus` function.
- Do we want default hotkeys for the new actions?

## Validation Steps Performed
At this point, compilation and manual testing of functionality (with hotkeys) by creating panes, adding distinguishers to each pane, and then swapping them around to confirm they went to the right location.

### CODE REVIEW

File: src/cascadia/TerminalSettingsModel/defaults.json
Comment (zadjii-msft) [inline src/cascadia/TerminalSettingsModel/defaults.json:351]: I'd take these out of the default bindings for now. People really really dislike us camping keys like these. You could theoretically leave them here, but without `keys` set, so they still appear in the command palette.

Comment (zadjii-msft) [reply src/cascadia/TerminalSettingsModel/defaults.json:351]: e.g.
```suggestion
        { "command": { "action": "movePane", "direction": "down" } },
        { "command": { "action": "movePane", "direction": "left" } },
        { "command": { "action": "movePane", "direction": "right" } },
        { "command": { "action": "movePane", "direction": "up" } },
        { "command": { "action": "movePane", "direction": "previous"} },
```

**Review by zadjii-msft — CHANGES_REQUESTED** [review]
Okay I haven't looked at Pane.cpp yet, but from a cursory glance, everything else here looks good. I'll probably come back through to actually read Pane.cpp sometime later this week. 

Very excited about this!

**Comment by Rosefield** [comment]
I attempted to add some basic test cases for the move panes functionality, but getting them ran has so far eluded me.
In the visual studio test runner, and when trying to run `\bin\x64\Debug\LocalTests_TerminalApp\TE.exe .\TerminalApp.LocalTests.dll` both fail with messages akin to
```
Error: TAEF: [HRESULT: 0x80070002] Failed to create the test host process for out of process test execution. (A custom AppxManifest does not exist at [snip]\terminal\bin\x64\Debug\LocalTests_TerminalApp\TestHostAppXManifest.xml.)
```

The `TestHostApp` project also refuses to build for me, with a large number of errors about not being able to instantiate abstract `Monarch` and `Peasant` classes in `RemotingTests.cpp`

**Comment by zadjii-msft** [comment]
Yea, you'll definitely need to build `TestHostApp` first to be able to run the tests.

> The `TestHostApp` project also refuses to build for me, with a large number of errors about not being able to instantiate abstract `Monarch` and `Peasant` classes in `RemotingTests.cpp`

That's definitely not expected, what errors exactly are you seeing? I could have swore that `TestHostApp` didn't depend on the RemotingTests at all...

**Comment by Rosefield** [comment]
Taking a look again, TestHostApp does build, and I did check that it does not depend on RemotingTests.cpp. I may have done a full solution build when I saw that problem? Regardless I just noticed the powershell Invoke-OpenConsoleTests that does seem to work for me so I'll try that out.

File: src/cascadia/LocalTests_TerminalApp/TabTests.cpp
Comment (Rosefield) [inline src/cascadia/LocalTests_TerminalApp/TabTests.cpp:894]: I am having a bit of troubles with this test with (presumably) threading issues. Specifically when running the test, if I attach a debugger and step through everything works correctly, however if I just run it normally it will fail here and the pane's children are the same before and after `_HandleMovePane`. (Yes I know that the IDs are incorrect here right now, I was going to wait to commit again until I figure out getting the test to pass consistently)

**Comment by ianjoneill** [comment]
I was also seeing issues when compiling the `UnitTests_Remoting` module, and thought it was a configuration issue on my machine.

It seems to have been fixed by #10593, so if you sync your fork the compilation issues should go away.

**Comment by Rosefield** [comment]
Assuming this gets approved, I took a quick peak at doing #2398 as well, and the commit is rather simple https://github.com/Rosefield/terminal/commit/b6da1ead21b9aac2794d8ecec6fe11320e39d2fd.

File: src/cascadia/TerminalApp/Pane.h
Comment (zadjii-msft) [inline src/cascadia/TerminalApp/Pane.h:230]: okay, this is gonna seem like a nit: Could we rename this member to `current`? I think that might help the readability, in places like

```c++
        if (_IsAdjacent(focus.focus, focus.focusOffset, shared_from_this(), offset, direction))
```

That would become

```
        if (_IsAdjacent(focus.current, focus.focusOffset, shared_from_this(), offset, direction))
```

File: src/cascadia/TerminalApp/Pane.h
Comment (zadjii-msft) [inline src/cascadia/TerminalApp/Pane.h:154]: Other nit: could we rename this to `_FindNeighborForPane`? The two methods added here (`_FindNeighborFromFocus` and `_FindFocusAndNeighbor`) are pretty similar name-wise. More distinct names would help the reader parse what the code's trying to do here.

File: src/cascadia/LocalTests_TerminalApp/TabTests.cpp
Comment (zadjii-msft) [reply src/cascadia/LocalTests_TerminalApp/TabTests.cpp:894]: TBH, all these UI like tests have those sorts of issues. If it were me, I'd just split this up into more `RunOnUIThread` calls, and maybe even insert some `Sleep`s between 😄 

I'd probably split this up.... (see above)

File: src/cascadia/LocalTests_TerminalApp/TabTests.cpp
Comment (zadjii-msft) [inline src/cascadia/LocalTests_TerminalApp/TabTests.cpp:888]:
```suggestion
            page->_HandleMovePane(nullptr, eventArgs);
        }
        VERIFY_SUCCEEDED(result);
                
        result = RunOnUIThread([&page]() {    
            auto tab = page->_GetTerminalTabImpl(page->_tabs.GetAt(0));
```

Comment (zadjii-msft) [reply src/cascadia/LocalTests_TerminalApp/TabTests.cpp:888]: Splitting up UI calls like this gives the UI thread a chance to settle itself after modifying the UI layout.

**Review by zadjii-msft — APPROVED** [review]
Okay, I'm cool with this. I have nits, but I wouldn't dare call any of them blocking. This is really cool stuff, excited to see the follow ups land too!

File: src/cascadia/TerminalApp/Pane.h
Comment (Rosefield) [reply src/cascadia/TerminalApp/Pane.h:230]: Maybe this particular example would be improved be renaming the local variable to something like `searchResult`? That way we have code that looks like
```        if (_IsAdjacent(searchResult.focus, searchResult.focusOffset, shared_from_this(), offset, direction))```
I feel like `current` might also cause confusion with the current pane i.e. `this`.

Comment (zadjii-msft) [reply src/cascadia/TerminalApp/Pane.h:230]: That's a good idea

File: src/cascadia/LocalTests_TerminalApp/TabTests.cpp
Comment (Rosefield) [reply src/cascadia/LocalTests_TerminalApp/TabTests.cpp]: I feel dirty adding as many sleeps as I did (they may not all be necessary), but at least the test passes consistently now.

**Comment by Rosefield** [comment]
@zadjii-msft What are the next steps to getting this merged? Thanks again for all of your time for reviewing.

**Comment by zadjii-msft** [comment]
At this point we need another signoff - @PankajBhojwani and @DHowett are fairly well versed in this part of the codebase, but it could be anyone on the team IMO.

File: src/cascadia/TerminalApp/Pane.cpp
Comment (lhecker) [inline src/cascadia/TerminalApp/Pane.cpp:584]: If we view this problem from the point of the focused pane, what this function (and the other) do is basically:
* move to the parent node
* check if the _other_ child node - the one the focused pane doesn't belong to - is in the direction we want to move the pane to
  * if yes: swap panes
  * if no: continue with the parent node

Is that correct?
If my understanding is correct, I believe we can simplify this code after it has been merged.

**Review by lhecker — CHANGES_REQUESTED** [review]
I'm not "requesting changes" because I'd like you to change anything.
Rather I'd like to simply discuss some of the points I brought up before we merge this.

File: src/cascadia/TerminalApp/TerminalTab.cpp
Comment (Rosefield) [reply src/cascadia/TerminalApp/TerminalTab.cpp:512]: I think you're probably right that this is unsafe, I can add a check to both of these functions.

File: src/cascadia/TerminalApp/Pane.cpp
Comment (Rosefield) [reply src/cascadia/TerminalApp/Pane.cpp:330]: I did it this way to avoid having yet another recursive helper function, but also it is somewhat vestigial on when I captured additional state and tried to modify it in place. I can see about refactoring a bit.

File: src/cascadia/TerminalApp/Pane.cpp
Comment (Rosefield) [reply src/cascadia/TerminalApp/Pane.cpp:2032]: I did it this way because having the conjunction of predicates means that a pane that is a leaf, but doesn't have the right ID would go into the `else` block, but it doesn't have children and so it would crash.

File: src/cascadia/TerminalSettingsModel/ActionArgs.idl
Comment (zadjii-msft) [reply src/cascadia/TerminalSettingsModel/ActionArgs.idl:165]: imo I'd rather keep the args for various different actions all separate, even if they coincidentally end up with the same properties. I think that helps keep the usages of these classes more specifc. You don't risk accidentally adding properties to another action when modifying one. Like, if we did ever want to modify one of these two actions in the future, keeping them separate now would be easier. Especially considering the `movePane(from=1, to=3)` case, where we're swapping panes by their indicies, and not just a direction

File: src/cascadia/TerminalApp/Pane.cpp
Comment (Rosefield) [reply src/cascadia/TerminalApp/Pane.cpp:584]: That is mostly a correct summary, yes. I do notice that my comment on this function is out of date so I'll update that.

File: src/cascadia/TerminalApp/AppCommandlineArgs.cpp
Comment (Rosefield) [reply src/cascadia/TerminalApp/AppCommandlineArgs.cpp:409]: This absolutely was just copied from MoveFocusParser. However, without using `#define`s I'm not sure how it could be abstracted because it relies on class members and types that are different for each. It isn't much of an abstraction if I have to pass in 3 different string arguments, pointers to some locals, and also have it parameterized by the argument type.

File: src/cascadia/TerminalApp/Pane.cpp
Comment (Rosefield) [reply src/cascadia/TerminalApp/Pane.cpp]: I created a `_FindParentOfPane` function and used it to replace the the vector-based logic.

File: src/cascadia/TerminalApp/TerminalTab.cpp
Comment (Rosefield) [reply src/cascadia/TerminalApp/TerminalTab.cpp:520]: Added a check to return early if `_mruPanes.size() < 2`

File: src/cascadia/TerminalApp/Pane.cpp
Comment (zadjii-msft) [reply src/cascadia/TerminalApp/Pane.cpp:566]: We can file this as a follow up - it's not blocking this particular PR, but yea the code probably could be improved.

**Review by lhecker — APPROVED** [review]

**Pull request merged**
